support inlining of output files in the action result

buchgr opened the pull request:
Allow the client to request the inlining of action outputs,
stdout and stderr in the action result message. We only add
this capability to the GetActionResult call for now as its
typically on the latency sensitive fast path, while Execute
is not.

Also make the stdout and stderr inlining decisions client
controlled and recommend for the server to keep the serialized
size of the ActionResult message below 1 MiB [1].

See https://groups.google.com/d/msg/remote-execution-apis/bmUuI2wUcwA/p0yrYq7VBgAJ

[1] https://developers.google.com/protocol-buffers/docs/techniques#large-data

buchgr commented:
@werkt @ishikhman @EricBurnett @juergbi @edbaunton

edbaunton commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -731,21 +731,31 @@ message ExecutedActionMetadata {
+//
+// The server MUST NOT make any inlining decisions on his own but should
+// do so when requested by the client. The total serialized message size
+// SHOULD NOT exceed 1 MiB.
```
Where does this 1MiB limit come from? A gRPC message size limit?

buchgr replied:
See commit message :-). It's not a limit but advice from protobuf https://developers.google.com/protocol-buffers/docs/techniques#large-data

EricBurnett replied:
I'd prefer to drop this recommendation, personally - we use >1MB messages all over the place today, and have seen no issues with doing so. My read is also that it wasn't intended to be a hard limit - the linked doc reads more like guidance not to use 100MB messages than it does a concrete recommendation that e.g. 4MB would be too much.

gRPC itself provides an upper bound on the size of messages - 4MB by default, if I remember correctly. I don't think we need a documented recommendation beyond that.

EricBurnett commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -731,21 +731,31 @@ message ExecutedActionMetadata {
 // An ActionResult represents the result of an
 // [Action][build.bazel.remote.execution.v2.Action] being run.
+//
+// The server MUST NOT make any inlining decisions on his own but should
```
Can we remove this comment? It's documented on the fields below, which seems like the appropriate place for it; I'm not sure that message-level comment adds anything.

Also, the server is free to *not* inline, so this one as written seems overly strong.

buchgr replied:
done

EricBurnett commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -731,21 +731,31 @@ message ExecutedActionMetadata {
 
+  // The output files of the action whose contents are inlined into the request.
+  // A server MUST NOT inline outputs unless requested by the client in the
+  // [GetActionResultRequest][build.bazel.remote.execution.v2.GetActionResultRequest]
```
MUST NOT->SHOULD NOT?

I'm hesitant to make it a MUST, in that it'll be mostly benign if it happens - no correctness guarantees will be violated due to over-inlining, just some amount of bandwidth wastage. As such, SHOULD NOT seems more correct (from the descriptions in https://www.ietf.org/rfc/rfc2119.txt).

Same below.

buchgr replied:
Done, also below.

EricBurnett commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -841,12 +851,13 @@ message ActionResult {
+  // stdout unless requested by the client in the
+  // [GetActionResultRequest][build.bazel.remote.execution.v2.GetActionResultRequest]
+  // message and instead return a digest in `stdout_digest` that points to
+  // the buffer. The client SHOULD NOT assume it will get one of the raw
```
Somewhat orthogonal to your proposal, but I think we err'd in saying that stdout_raw may be set and stdout_digest then left empty. I'd prefer the semantics be that stdout_digest is always set, and stdout_raw may *also* be set if inlining is requested. Otherwise if a client both wants to display the stdout and save a reference to it for later, the client needs to *upload* it to the CAS for persistence themselves, which seems backwards.

If we're changing the semantics of inlining anyways, and no implementation we know of relies on the semantics of today, could we change this?

ulfjack replied:
I agree that it would be better to report the bytes and the digest.

buchgr replied:
Great idea. Done.

EricBurnett commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -894,6 +906,24 @@ message OutputFile {
 
+// `InlinedOutputFile` is similar to [OutputFile][build.bazel.remote.execution.v2.OutputFile]
+// but has its contents inlined in the message.
+message InlinedOutputFile {
```
Why not just add an optional 'contents' field to OutputFile? I'm not a huge fan of the duplication this currently requires, and don't really see why it would be necessary to segregate.

Tangential question: Anyone know why "`OutputFile` is binary-compatible with `FileNode`." ? I assume that was motivated by something in the past, but I can't figure out what. It's probably the wrong time to change that (should be a V3 feature request), but can still be satisfied by reserving the field number in FileNode, as is done with 3 now.

ola-rozenfeld replied:
+1 to that. I think we should just revert https://github.com/bazelbuild/remote-apis/pull/5 to start with.

I also can't remember for the life of me why OutputFile should be binary compatible with FileNode either. I don't think it should.

buchgr replied:
I created the extra message because of binary compatibility requirement with FileNode. I agree to adding the field to `OutputFile` instead.

EricBurnett commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -1126,6 +1156,19 @@ message GetActionResultRequest {
   Digest action_digest = 2;
+
+  // A hint to the server to inline stdout in the
+  // [ActionResult][build.bazel.remote.execution.v2.ActionResult] message.
```
inline -> request inlining ? I think that helps reinforce the message that the server may not comply.

buchgr replied:
done

buchgr commented:
Apologies for the delay in responding to comments. I have been quite busy with other things and all of a sudden a month had passed :-). I believe I have addressed all comments. I ll squash and update the commit message after getting an LGTM.

ola-rozenfeld commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -841,32 +841,28 @@ message ActionResult {
+  // stderr unless requested by the client in the
+  // [GetActionResultRequest][build.bazel.remote.execution.v2.GetActionResultRequest]
+  // message and instead return a digest in `stderr_digest` that points to
+  // the buffer.
```
The comments still make it look like the digest is optional.

buchgr replied:
done.

ola-rozenfeld commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -892,6 +886,11 @@ message OutputFile {
+  // The contents of the file if inlining was requested. The server SHOULD NOT inline
+  // file contents unless requested by the client in the
+  // [GetActionResultRequest][build.bazel.remote.execution.v2.GetActionResultRequest]
+  bytes contents = 5;
```
I know that usually we don't reuse removed tags, but this is a rare case of the exact same field being removed and then ressurrected. I think this justifies using bytes contents = 3, removing the "reserved".

buchgr replied:
Meh. I'd prefer to stick to the proto guidelines if you don't mind.

EricBurnett commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -892,6 +884,11 @@ message OutputFile {
   // True if file is executable, false otherwise.
   bool is_executable = 4;
+
+  // The contents of the file if inlining was requested. The server SHOULD NOT inline
```
Should we be explicit somewhere that the server MAY omit inlining, even if requested, and MUST do so if inlining would cause the response to exceed message size limits? Possibly here (and on other output fields), possibly on the relevant fields in GetActionResultRequest instead.

buchgr replied:
done.

EricBurnett commented on build/bazel/remote/execution/v2/remote_execution.proto:
```diff
@@ -737,8 +737,8 @@ message ActionResult {
   // the action completed, a single entry will be present either in this field,
-  // or in the output_file_symlinks field, if the file was a symbolic link to
-  // another file.
+  // or the `inlined_output_files` field or the `output_file_symlinks` field if the
```
Out of date? inlined_output_files no longer exists, I don't think.

buchgr replied:
done.

buchgr commented:
Good to go?

ishikhman commented:
LGTM ;)

EricBurnett commented:
LGTM as well, thanks Jakob.
